doc/website/blog/2021-09-02-ent-extension-api.md

```diff
@@ -0,0 +1,280 @@
+and how that made the Protobuf ecosystem thrive. In short, ecosystems are only created under
+modular designs. 
+
+In our post today, we will explore Ent's `Extension` API by building a toy example. 
+
+### Getting Started
+
+The Extension API only works for projects use Ent's code-generation [as a Go package](https://entgo.io/docs/code-gen/#use-entc-as-a-package).
+To set that up, after initializing your project, create a new file named `ent/entc.go`:
+```go title=ent/entc.go
+//+build ignore
+
+package main
+
+import (
+    "log"
+
+    "entgo.io/ent/entc"
+    "entgo.io/ent/entc/gen"
+    "entgo.io/ent/schema/field"
+)
+
+func main() {
+    err := entc.Generate("./schema", &gen.Config{})
+    if err != nil {
+        log.Fatal("running ent codegen:", err)
+    }
+}
+```
+Next, modify `ent/generate.go` to invoke our `entc` file:
+```go title=ent/generate.go
+package ent
+
+//go:generate go run entc.go
+```
+
+### Creating our Extension
+
+All extension's must implement the [Extension](https://pkg.go.dev/entgo.io/ent/entc#Extension) interface:
+
+```go
+type Extension interface {
+	// Hooks holds an optional list of Hooks to apply
+	// on the graph before/after the code-generation.
+	Hooks() []gen.Hook
+	// Annotations injects global annotations to the gen.Config object that
+	// can be accessed globally in all templates. Unlike schema annotations,
+	// being serializable to JSON raw value is not mandatory.
+	//
+	//	{{- with $.Config.Annotations.GQL }}
+	//		{{/* Annotation usage goes here. */}}
+	//	{{- end }}
+	//
+	Annotations() []Annotation
+	// Templates specifies a list of alternative templates
+	// to execute or to override the default.
+	Templates() []*gen.Template
+	// Options specifies a list of entc.Options to evaluate on
+	// the gen.Config before executing the code generation.
+	Options() []Option
+}
+```
+To simplify the development of new extensions, developers can embed [entc.DefaultExtension](https://pkg.go.dev/entgo.io/ent/entc#DefaultExtension)
+to create extensions without implementing all methods. In `entc.go`, add: 
+```go title=ent/entc.go
+// ...
+
+// GreetExtension implements entc.Extension.
+type GreetExtension {
+	entc.DefaultExtension
+}
+```
+
+Currently, our extension doesn't do anything. Next, let's connect it to our code-generation config.
+In `entc.go`, add our new extension to the `entc.Generate` invocation:
+
+```go
+err := entc.Generate("./schema", &gen.Config{}, entc.Extensions(&GreetExtension{})
+```
+
+### Adding Templates
+
+External templates can be bundled into extensions to enhance Ent's core code-generation
+functionality. With our toy example, our goal is to add to each entity a generated method
+name `Greet` that returns a greeting with the type's name when invoked. We're aiming for something 
+like:
+
+```go
+func (u *User) Greet() string {
+    return "Greetings, User"
+}
+```
+
+To do this, let's add a new external template file and place it in `ent/templates/greet.tmpl`:
```

Comment:
Maybe worth letting the reader understand where he/she can understand template language from.